beacon-chain/sync/block_batcher.go

```diff
@@ -0,0 +1,190 @@
+
+	nb.seq, nb.nonseq, nb.err = filterCanonical(ctx, rob, &bb.lastSeq, bb.isCanonical)
+
+	// Decrease allowed blocks capacity by the number of streamed blocks.
+	bb.limiter.add(stream, int64(1+nb.end.SubSlot(nb.start)))
+	bb.current = &nb
+	return *bb.current, true
+}
+
+func newBlockBatch(start, reqEnd primitives.Slot, size uint64) (blockBatch, bool) {
+	if start > reqEnd {
+		return blockBatch{}, false
+	}
+	nb := blockBatch{start: start, end: start.Add(size - 1)}
+	if nb.end > reqEnd {
+		nb.end = reqEnd
+	}
+	return nb, true
+}
+
+type blockBatch struct {
+	start  primitives.Slot
+	end    primitives.Slot
+	seq    []blocks.ROBlock
+	nonseq []blocks.ROBlock
+	err    error
+}
+
+// sequence returns the list of sequential, canonical blocks read from the db.
+func (bb blockBatch) sequence() []blocks.ROBlock {
+	return bb.seq
+}
+
+// sequenceBroken is used to determine if there was a break in the chain of canonical blocks as read from the db.
+// If true, code using the blockBatch should stop serving additional batches of blocks.
+func (bb blockBatch) sequenceBroken() bool {
+	return len(bb.nonseq) > 0
+}
+
+func (bb blockBatch) error() error {
+	return bb.err
+}
+
+type canonicalChecker func(context.Context, [32]byte) (bool, error)
+
+// filters all the provided blocks to ensure they are canonical
+// and are strictly linear.
+func filterCanonical(ctx context.Context, blks []blocks.ROBlock, prevRoot *[32]byte, canonical canonicalChecker) ([]blocks.ROBlock, []blocks.ROBlock, error) {
+	seq := make([]blocks.ROBlock, 0, len(blks))
+	nseq := make([]blocks.ROBlock, 0)
+	for i, b := range blks {
+		cb, err := canonical(ctx, b.Root())
+		if err != nil {
+			return nil, nil, err
+		}
+		if !cb {
+			continue
+		}
+		// filterCanonical is called in batches, so prevRoot can be the last root from the previous batch.
+		// prevRoot will be the zero value until we find the first canonical block in a given request.
+		first := *prevRoot == [32]byte{}
+		// We assume blocks are processed in order, so the previous canonical root should be the parent of the next.
+		// If the current block isn't descended from the last, something is wrong. Append everything remaining
+		// to the list of non-sequential blocks and stop building the canonical list.
+		if !first && *prevRoot != b.Block().ParentRoot() {
+			nseq = append(nseq, blks[i:]...)
+			break
+		}
+		seq = append(seq, blks[i])
+		// Set the previous root as the
+		// newly added block's root
+		currRoot := b.Root()
+		*prevRoot = currRoot
+	}
+	return seq, nseq, nil
+}
+
+// returns a copy of the []ROBlock list in sorted order with duplicates removed
+func sortedUniqueBlocks(blks []blocks.ROBlock) []blocks.ROBlock {
+	// Remove duplicate blocks received
+	sort.Sort(blocks.ROBlockSlice(blks))
+	if len(blks) < 2 {
+		return blks
+	}
+	u := 0
+	for i := 1; i < len(blks); i++ {
+		if blks[i].Root() != blks[u].Root() {
+			u += 1
+			if u != i {
+				blks[u] = blks[i]
+			}
+		}
+	}
+	return blks[:u+1]
```

Comment:
Note that I didn't just move this, I rewrote it to take advantage of the sorted invariant.